crates/query-engine/translation/src/translation/query/values.rs

```diff
@@ -84,48 +84,88 @@ fn type_to_ast_scalar_type(typ: &database::Type) -> sql::ast::ScalarType {
-                    },
-                    exp,
-                ],
             }
         }
-        _ => {
-            let exp = sql::ast::Expression::BinaryOperation {
-                left: Box::new(variables_reference),
-                operator: sql::ast::BinaryOperator("->>".to_string()),
-                right: Box::new(sql::ast::Expression::Value(sql::ast::Value::String(
-                    variable.clone(),
-                ))),
+        database::Type::ScalarType(_) => sql::ast::Expression::BinaryOperation {
+            left: Box::new(variables_reference),
+            operator: sql::ast::BinaryOperator("->>".to_string()),
+            right: Box::new(sql::ast::Expression::Value(sql::ast::Value::String(
+                variable.clone(),
+            ))),
+        },
+    };
+
+    translate_projected_variable(state, r#type, projected_variable_exp)
+}
+
+pub fn translate_projected_variable(
+    state: &mut State,
+    r#type: &database::Type,
+    exp: sql::ast::Expression,
+) -> sql::ast::Expression {
+    match r#type {
+        database::Type::CompositeType(_type_name) => sql::ast::Expression::FunctionCall {
+            function: sql::ast::Function::JsonbPopulateRecord,
+            args: vec![
+                sql::ast::Expression::Cast {
+                    expression: Box::new(sql::ast::Expression::Value(sql::ast::Value::Null)),
+                    r#type: type_to_ast_scalar_type(r#type),
+                },
+                exp,
+            ],
+        },
+        database::Type::ArrayType(type_name) => {
+            let arr_table = state.make_table_alias("arr".to_string());
+            let elem_column = sql::ast::ColumnAlias {
+                name: "elem".to_string(),
             };
-            sql::ast::Expression::Cast {
-                expression: Box::new(exp),
-                r#type: type_to_ast_scalar_type(r#type),
-            }
+
+            let from_arr = sql::ast::From::JsonbArrayElements {
+                expression: exp,
+                alias: arr_table.clone(),
+                column: elem_column.clone(),
+            };
+
+            let elem_exp = sql::ast::Expression::ColumnReference(ColumnReference::AliasedColumn {
+                table: sql::ast::TableReference::AliasedTable(arr_table.clone()),
+                column: elem_column.clone(),
+            });
+
+            let converted_element_exp = translate_projected_variable(state, type_name, elem_exp);
+
+            let mut result_select = simple_select(vec![(
+                elem_column.clone(),
+                sql::ast::Expression::FunctionCall {
+                    function: sql::ast::Function::Unknown("array_agg".to_string()),
+                    args: vec![converted_element_exp],
+                },
+            )]);
+
+            result_select.from = Some(from_arr);
+
+            sql::ast::Expression::CorrelatedSubSelect(Box::new(result_select))
         }
+        _ => sql::ast::Expression::Cast {
```

Comment:
If we can avoid catch-all patterns, let's do that.

```suggestion
        database::Type::ScalarType(_type_name) => sql::ast::Expression::Cast {
```

Reply:
done.